pallets/delegating/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+                score += s;
+            }
+            Some(score)
+        } else {
+            Some(0)
+        }
+    }
+
+    fn total_delegated_score() -> Option<u64> {
+        // check delegators credit score
+        Self::check_and_adjust_delegated_score();
+
+        let mut total_score: u64 = 0;
+        if let Some(candidate_validators) = <CandidateValidators<T>>::get() {
+            for candidate_validator in candidate_validators {
+                total_score = total_score.saturating_add(
+                    Self::delegated_score_of_validator(&candidate_validator).unwrap_or(0),
+                );
+            }
+        }
+        Some(total_score)
+    }
+
+    fn get_total_validator_score(era_index: EraIndex, validator: T::AccountId) -> Option<u64> {
+        if <SelectedDelegators<T>>::contains_key(era_index, validator.clone()) {
+            let delegators = <SelectedDelegators<T>>::get(era_index, validator);
+            let mut total_score: u64 = 0;
+            for (_, s, _) in delegators {
+                total_score += s;
+            }
+            Some(total_score)
+        } else {
+            Some(0)
+        }
+    }
+
+    fn set_eras_reward(era_index: EraIndex, total_reward: BalanceOf<T>) {
+        <ErasValidatorReward<T>>::insert(era_index, total_reward);
+    }
+
+    fn payout_delegators(
+        era_index: EraIndex,
+        commission: Perbill,
+        validator_reward_part: Perbill,
+        validator: T::AccountId,
+        validator_payee: T::AccountId,
+    ) -> bool {
+        if !<ErasValidatorReward<T>>::contains_key(&era_index) {
+            return false;
+        }
+        let total_payout = <ErasValidatorReward<T>>::get(&era_index).unwrap();
+        let era_payout = validator_reward_part * total_payout;
+
+        let validator_commission_payout = commission * era_payout;
+        let validator_leftover_payout = era_payout - validator_commission_payout;
+
+        // We can now make total validator payout:
+        if let Some(imbalance) =
+            Self::make_payout(validator_payee.clone(), validator_commission_payout)
+        {
+            Self::deposit_event(RawEvent::Reward(validator_payee, imbalance.peek()));
+        }
+
+        // Lets now calculate how this is split to the nominators.
+        // Reward only the clipped exposures. Note this is not necessarily sorted.
+        let era_total_score =
+            Self::get_total_validator_score(era_index, validator.clone()).unwrap();
+        let delegators = <SelectedDelegators<T>>::get(era_index, validator.clone());
+        for (who, s, _) in delegators {
+            let delegator_exposure_part = Perbill::from_rational_approximation(s, era_total_score);
+
+            let delegator_reward: BalanceOf<T> =
+                delegator_exposure_part * validator_leftover_payout;
+            // We can now make nominator payout:
+            if let Some(imbalance) = Self::make_payout(who.clone(), delegator_reward) {
```

Comment:
when an imbalance will be created? Is it just simply dropped if imbalance occurs? If simply dropped, can we put it into the treasury pool instead?

Reply:
We try to using "T::Currency::deposit_into_existing(stash, amount).ok()"